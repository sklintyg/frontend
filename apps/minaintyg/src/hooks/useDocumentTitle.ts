/* eslint-disable react-hooks/exhaustive-deps */
import { RefObject, useEffect } from 'react'

function updateDocumentTitle(el: HTMLElement | null) {
  if (el && el.innerText && el.innerText.length > 0) {
    document.title = `${el.innerText} - 1177`
  }
}

export function useDocumentTitle(ref: RefObject<HTMLElement>, dependencies: unknown[]) {
  useEffect(() => {
    const observer = new MutationObserver(() => {
      updateDocumentTitle(ref.current)
    })

    if (ref.current) {
      observer.observe(ref.current, { childList: true })
      updateDocumentTitle(ref.current)
    }
    return () => {
      observer.disconnect()
    }
  }, dependencies)
}
