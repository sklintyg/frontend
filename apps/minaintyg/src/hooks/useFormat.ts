import { format, parseISO } from 'date-fns'

export function useFormat() {
  const formatDate = (timestamp: string) => format(parseISO(timestamp), 'yyyy-MM-dd')
  const formatTime = (timestamp: string) => format(parseISO(timestamp), 'HH:mm')

  return {
    formatDate,
    formatTime,
    formatDatetime: (timestamp: string) => `${formatDate(timestamp)} ${formatTime(timestamp)}`,
  }
}
