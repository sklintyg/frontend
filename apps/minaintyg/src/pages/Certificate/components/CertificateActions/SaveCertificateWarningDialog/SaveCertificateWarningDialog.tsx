import { Dialog } from '@frontend/components'
import { ReactNode } from 'react'

export function SaveCertificateWarningDialog({
  open,
  onOpenChange,
  children,
}: {
  open: boolean
  onOpenChange?: (open: boolean) => void
  children: ReactNode
}) {
  return (
    <Dialog open={open} onOpenChange={onOpenChange} headline="Spara intyg som PDF">
      <p className="mb-5">Viktigt! Ditt intyg har information om dig, som ditt namn och andra personuppgifter.</p>
      <p>
        När du sparar intyget som en PDF-fil, sparas filen på den enhet du använder. Om du använder en dator som andra också kan använda,
        till exempel på ett bibliotek, måste du själv ta bort filen från datorn innan du går därifrån.
      </p>
      {children}
    </Dialog>
  )
}
