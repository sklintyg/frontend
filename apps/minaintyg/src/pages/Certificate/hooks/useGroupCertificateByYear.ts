import { sortBy } from '@frontend/utils'
import { getYear, parseISO } from 'date-fns'
import { useMemo } from 'react'
import { SortDirection } from 'react-stately'
import { CertificateMetadata } from '../../../schema/certificate.schema'

const getCertificateYear = ({ issued }: CertificateMetadata) => getYear(parseISO(issued))

export function useGroupCertificateByYear(certificates: CertificateMetadata[], order: SortDirection = 'descending') {
  return useMemo(() => {
    // Could be replaced with Object.groupBy in the future (currently experimental)
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy#browser_compatibility
    const groupByYear = [...certificates]
      .sort(sortBy(order, ({ issued }) => new Date(issued).getTime()))
      .reduce<Record<string, CertificateMetadata[]>>((result, certificate) => {
        const year = getCertificateYear(certificate)
        return { ...result, [year]: [...(result[year] ?? []), certificate] }
      }, {})

    return Object.entries(groupByYear).sort(sortBy(order, ([year]) => parseInt(year, 10)))
  }, [certificates, order])
}
