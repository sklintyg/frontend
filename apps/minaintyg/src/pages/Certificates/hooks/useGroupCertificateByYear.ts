import { sortBy } from '@frontend/utils'
import { getYear, parseISO } from 'date-fns'
import { useMemo } from 'react'
import { SortDirection } from 'react-stately'
import { CertificateListItem } from '../../../schema/certificateList.schema'

const getCertificateYear = ({ issued }: CertificateListItem) => getYear(parseISO(issued))

export function useGroupCertificateByYear(certificates: CertificateListItem[], order: SortDirection = 'descending') {
  return useMemo(() => {
    // Could be replaced with Object.groupBy in the future (currently experimental)
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy#browser_compatibility
    const groupByYear = [...certificates]
      .sort(sortBy(order, ({ issued }) => new Date(issued).getTime()))
      .reduce<Record<string, CertificateListItem[]>>((result, certificate) => {
        const year = getCertificateYear(certificate)
        return { ...result, [year]: [...(result[year] ?? []), certificate] }
      }, {})

    return Object.entries(groupByYear).sort(sortBy(order, ([year]) => parseInt(year, 10)))
  }, [certificates, order])
}
