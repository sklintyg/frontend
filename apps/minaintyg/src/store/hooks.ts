import { skipToken } from '@reduxjs/toolkit/query'
import type { TypedUseSelectorHook } from 'react-redux'
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import { shallowEqual, useDispatch, useSelector } from 'react-redux'
import { useLocation } from 'react-router-dom'
import type { AvailableFunctionsType } from '../schema/certificate.schema'
import { api } from './api'
import type { RootState } from './reducer'
import type { AppDispatch } from './store'

// Use throughout your app instead of plain `useDispatch` and `useSelector`
type DispatchFunc = () => AppDispatch
export const useAppDispatch: DispatchFunc = useDispatch
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

export function useGetUserQuery() {
  const location = useLocation()
  const hasSessionEnded = useAppSelector((state) => state.sessionSlice.hasSessionEnded)
  const isErrorPage = location.pathname.startsWith('/error')

  return api.useGetUserQuery(hasSessionEnded || isErrorPage ? skipToken : undefined)
}

export function useAvailableFunction(id: string, type: AvailableFunctionsType) {
  return useAppSelector((state) => {
    const certificate = api.endpoints.getCertificate.select({ id })(state).data
    return (certificate?.availableFunctions ?? []).find((fn) => fn.type === type)
  }, shallowEqual)
}
