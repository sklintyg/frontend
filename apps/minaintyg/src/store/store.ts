import { configureStore } from '@reduxjs/toolkit'
import { api } from './api'
import { errorMiddleware } from './middleware/error.middleware'
import { reducer } from './reducer'
import { idpMiddleware } from './slice/idpMiddleware'

export const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([api.middleware, idpMiddleware, errorMiddleware]),
})

export type AppDispatch = typeof store.dispatch
