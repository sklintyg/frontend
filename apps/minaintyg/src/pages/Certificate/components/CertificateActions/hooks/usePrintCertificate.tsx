import { format } from 'date-fns'
import { createContext, useContext, useState } from 'react'
import { AvailableFunctionsTypeEnum, InformationTypeEnum } from '../../../../../schema/certificate.schema'
import { useAvailableFunction } from '../../../../../store/hooks'

type CustomizePrintType = 'print' | 'save' | null

export function usePrintCertificate(id: string) {
  const [customizeId, setCustomizeId] = useState('')
  const [customizePrintType, setCustomizePrintType] = useState<CustomizePrintType>(null)
  const [customizePrintDialogOpen, setShowCustomizePrintDialog] = useState(false)
  const printFunction = useAvailableFunction(id, AvailableFunctionsTypeEnum.enum.PRINT_CERTIFICATE)
  const customizePrintFunction = useAvailableFunction(id, AvailableFunctionsTypeEnum.enum.CUSTOMIZE_PRINT_CERTIFICATE)

  const activePrintFunction = printFunction ?? customizePrintFunction

  const fileName = activePrintFunction?.information.find((info) => info.type === InformationTypeEnum.enum.FILENAME)
  const url = `/api/certificate/${id}/pdf/${fileName?.text}_${format(Date.now(), 'yy-MM-dd_HHmm')}.pdf${
    customizeId ? `?customizationId=${customizeId}` : ''
  }`

  function showCustomizePrintDialog(type: CustomizePrintType) {
    setShowCustomizePrintDialog(true)
    setCustomizePrintType(type)
  }

  function hideCustomizePrintDialog() {
    setShowCustomizePrintDialog(false)
  }

  return {
    url,
    customizePrintFunction,
    printFunction,
    customizeId,
    setCustomizeId,
    customizePrintDialogOpen,
    showCustomizePrintDialog,
    hideCustomizePrintDialog,
    setShowCustomizePrintDialog,
    customizePrintType,
  }
}

export const PrintCertificateContext = createContext<ReturnType<typeof usePrintCertificate> | null>(null)
export const PrintCertificateContextProvider = PrintCertificateContext.Provider

export function usePrintCertificateContext() {
  const state = useContext(PrintCertificateContext)

  if (state === null) {
    throw Error('Should be used inside PrintCertificateContextProvider')
  }

  return state
}
