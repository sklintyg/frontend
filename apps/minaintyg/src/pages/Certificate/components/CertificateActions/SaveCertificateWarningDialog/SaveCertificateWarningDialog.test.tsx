import { render, screen } from '@testing-library/react'
import { ReactNode } from 'react'
import { Provider } from 'react-redux'
import { store } from '../../../../../store/store'
import { PrintCertificateContextProvider, usePrintCertificate } from '../hooks/usePrintCertificate'
import { SaveCertificateWarningDialog } from './SaveCertificateWarningDialog'

function ComponentWrapper({ value, children }: { value: Partial<ReturnType<typeof usePrintCertificate>>; children: ReactNode }) {
  const defaultValue = usePrintCertificate('id')
  return (
    <PrintCertificateContextProvider value={{ ...defaultValue, ...value, setSaveWarningDialogOpen: vi.fn() }}>
      {children}
    </PrintCertificateContextProvider>
  )
}

function renderComponent(value: Partial<ReturnType<typeof usePrintCertificate>>) {
  return render(
    <Provider store={store}>
      <ComponentWrapper value={value}>
        <SaveCertificateWarningDialog />
      </ComponentWrapper>
    </Provider>
  )
}

it('Should render as expected', () => {
  renderComponent({ saveWarningDialogOpen: true })
  expect(screen.getByRole('dialog', { name: 'Spara intyg som PDF' })).toBeInTheDocument()
  expect(screen.getByText(/viktigt! Ditt intyg har information om dig, som ditt namn och andra personuppgifter/i)).toBeInTheDocument()
  expect(screen.getByText(/när du sparar intyget som en PDF-fil, sparas filen på den enhet du använder/i)).toBeInTheDocument()
})
