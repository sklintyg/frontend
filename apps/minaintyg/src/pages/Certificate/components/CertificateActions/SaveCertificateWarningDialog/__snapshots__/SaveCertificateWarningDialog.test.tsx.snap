// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render as expected 1`] = `
<div>
  <div
    aria-hidden="true"
    data-focus-guard="true"
    data-focus-on-hidden="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  <div
    data-focus-lock-disabled="false"
  >
    <ids-dialog
      aria-labelledby=":r0:"
      role="dialog"
    >
      <h1
        class="ids-heading-1 ids-small"
        id=":r0:"
        slot="headline"
      >
        Spara intyg som PDF
      </h1>
      <p
        class="mb-5"
      >
        Viktigt! Ditt intyg har information om dig, som ditt namn och andra personuppgifter. 
      </p>
      <p>
        När du sparar intyget som en PDF-fil, sparas filen på den enhet du använder. Om du använder en dator som andra också kan använda, till exempel på ett bibliotek, måste du själv ta bort filen från datorn innan du går därifrån.
      </p>
      Test
    </ids-dialog>
  </div>
  <div
    aria-hidden="true"
    data-focus-guard="true"
    data-focus-on-hidden="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
</div>
`;
